File: examples/python/operations/parallel/completion-config.py
from aws_durable_execution_sdk_python import (
    BatchResult,
    DurableContext,
    durable_execution,
)
from aws_durable_execution_sdk_python.config import CompletionConfig, ParallelConfig


def source_a(ctx: DurableContext) -> str:
    return ctx.step(lambda _: "result from a", name="source-a")


def source_b(ctx: DurableContext) -> str:
    return ctx.step(lambda _: "result from b", name="source-b")


def source_c(ctx: DurableContext) -> str:
    return ctx.step(lambda _: "result from c", name="source-c")


@durable_execution
def handler(event: dict, context: DurableContext) -> str | None:
    result: BatchResult[str] = context.parallel(
        [source_a, source_b, source_c],
        name="race",
        config=ParallelConfig(completion_config=CompletionConfig.first_successful()),
    )
    results = result.get_results()
    return results[0] if results else None

File: examples/csharp/operations/parallel/completion-config.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;

public class CompletionConfigExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<object, string?>(Workflow, input, context);

    private async Task<string?> Workflow(object input, IDurableContext ctx)
    {
        var config = new ParallelConfig
        {
            // Complete as soon as one branch succeeds
            CompletionConfig = CompletionConfig.FirstSuccessful(),
        };

        IBatchResult<string> result = await ctx.ParallelAsync(
            new Func<IDurableContext, CancellationToken, Task<string>>[]
            {
                async (branch, ct) => await branch.StepAsync(
                    async (_, _) => "result from a", name: "source-a"),
                async (branch, ct) => await branch.StepAsync(
                    async (_, _) => "result from b", name: "source-b"),
                async (branch, ct) => await branch.StepAsync(
                    async (_, _) => "result from c", name: "source-c"),
            },
            name: "race",
            config: config);

        return result.GetResults().FirstOrDefault();
    }
}
